backend/internal/authn/authnservice.go

```diff
@@ -0,0 +1,322 @@
+func getSessionData(key string) (*sessionmodel.SessionData, *serviceerror.ServiceError) {
+	sessionDataStore := sessionstore.GetSessionDataStore()
+	ok, sessionData := sessionDataStore.GetSession(key)
+	if !ok {
+		return nil, &constants.ErrorSessionNotFound
+	}
+
+	return &sessionData, nil
+}
+
+// handleFlowError handles errors that occur during flow execution and returns an appropriate authentication error.
+func handleFlowError(flowErr *serviceerror.ServiceError, logger *log.Logger) *serviceerror.ServiceError {
+	var retErr *serviceerror.ServiceError
+	if flowErr.Type == serviceerror.ClientErrorType {
+		retErr = &constants.ErrorFlowExecutionClientError
+	} else {
+		logger.Error("An error occurred while executing the flow", log.Any("error", flowErr))
+		retErr = &constants.ErrorFlowExecutionServerError
+	}
+	retErr.ErrorDescription = flowErr.ErrorDescription
+	return retErr
+}
+
+// handleFlowResponse processes the flow step and updates the session data store accordingly.
+// It returns an AuthNResponse if the flow is complete, or an error if any issues arise during processing.
+func handleFlowResponse(sessionDataKey, requestFlowID string, flowStep *flowmodel.FlowStep,
+	sessionData *sessionmodel.SessionData, logger *log.Logger) (*model.AuthNResponse, *serviceerror.ServiceError) {
+	sessionDataStore := sessionstore.GetSessionDataStore()
+	if isNewFlow(sessionDataKey, requestFlowID) {
+		if flowStep.Status == flowconst.FlowStatusIncomplete {
+			logger.Debug("Flow execution is incomplete, storing session data", log.String("flowID", flowStep.FlowID))
+			sessionDataStore.AddSession(flowStep.FlowID, *sessionData)
+		}
+	} else {
+		if flowStep.Status == flowconst.FlowStatusComplete {
+			var svcErr *serviceerror.ServiceError
+			sessionData, svcErr = getSessionData(flowStep.FlowID)
+			if svcErr != nil {
+				return nil, svcErr
+			}
+		}
+		if flowStep.Status != flowconst.FlowStatusIncomplete {
+			sessionDataStore.ClearSession(flowStep.FlowID)
+		}
+	}
+
+	if flowStep.Status == flowconst.FlowStatusComplete {
+		logger.Debug("Flow execution completed successfully", log.String("flowID", flowStep.FlowID))
+		return handleCompletedFlow(flowStep, sessionData)
+	}
+
+	return &model.AuthNResponse{
+		FlowID:        flowStep.FlowID,
+		StepID:        flowStep.StepID,
+		FlowStatus:    string(flowStep.Status),
+		Type:          string(flowStep.Type),
+		Data:          flowStep.Data,
+		FailureReason: flowStep.FailureReason,
+	}, nil
+}
+
+// handleCompletedFlow handles the flow step when the authentication flow is completed.
+// It decodes the user attributes from the flow assertion and constructs a new session data object.
+func handleCompletedFlow(flowStep *flowmodel.FlowStep, sessionData *sessionmodel.SessionData) (
+	*model.AuthNResponse, *serviceerror.ServiceError) {
+	userID, userAttributes, err := decodeAttributesFromAssertion(flowStep.Assertion)
+	if err != nil {
```

Comment:
Handling session data done at various levels..
- In the execute() level
- In the handleFlowResponse() level
- In the handleCompletedFlow() level
Can we refactor the code to handle session data in one layer..?